Add a readme for k8s autodiscover provider (#28213)
elastic/beats: libbeat/autodiscover/README.md

**Providers**
Autodiscover providers work by watching for events on the system and translating those events into internal autodiscover events with a common format. When you configure the provider, you can optionally use fields from the autodiscover event to set conditions that, when met, launch specific configurations.

On start, Metricbeat/Filebeat/Agent will scan existing containers and launch the proper configs for them. Then it will watch for new start/stop events. This ensures you don’t need to worry about state, but only define your desired configs.

The Kubernetes autodiscover provider watches for Kubernetes nodes, pods, services and namespaces when they start, update, and stop.

We will describe the internals of three ways of Kubernetes autodiscover process.
1. Templates Based Autodiscover
2. Hints Based Autodiscover
3. Autodiscover provider in Elastic Agent

## Templates Based Autodiscover

As the name suggests, user needs to set a template to indicate to autodiscover provider what to do.
There is one configuration variable that differentiates in a way how the autosicover process is performed.
This variable is `unique`

### Autodiscover with LeaderElection

> When setting `unique: true` the Leader Election mechanism is activated. That way **only** the beat process that will gain the leader lease will enable the provided template.

Reviewer: ```suggestion
When setting `unique: true` the Leader Election mechanism is activated. That way **only** the Beat instance that will gain the leader lease/lock will enable the provided template.
```

The best appliance of this feature is when collecting cluster wide metrics from `kube-state-metrics` or `apiserver`.
In that case having all instances of metricbeat collecting the same metrics is not desirable.

Example:
```
metricbeat.autodiscover:
  providers:
    - type: kubernetes
      scope: cluster
      node: ${NODE_NAME}
      unique: true
      identifier: leader-election-metricbeat
      templates:
        - config:
            - module: kubernetes
              hosts: ["kube-state-metrics:8080"]
              period: 10s
              add_metadata: true
              metricsets:
                - state_node
```

##### How it works

We will deep dive in the internals of [libbeat kubernetes autodiscover provider](https://github.com/elastic/beats/tree/master/libbeat/autodiscover/providers/kubernetes).

We will use metricbeat as an example.

Step-by-step walkthrough

> 1. Kubernetes provider `init` function [adds](https://github.com/elastic/beats/blob/master/libbeat/autodiscover/providers/kubernetes/kubernetes.go#L46) the provider in the autodiscover providers registry at startup. For Kubernetes provider a `AutodiscoverBuilder` func is passed as an argument.

Reviewer: ```suggestion
1. Kubernetes provider `init` function [adds](https://github.com/elastic/beats/blob/master/libbeat/autodiscover/providers/kubernetes/kubernetes.go#L46) the provider in the autodiscover providers registry at startup. For Kubernetes provider an `AutodiscoverBuilder` func is passed as an argument.
```

> 2. Metricbeat calls `NewAutodiscover` [function](https://github.com/elastic/beats/blob/master/metricbeat/beater/metricbeat.go#L183) which checks in the config for enabled providers and [builds](https://github.com/elastic/beats/blob/master/libbeat/autodiscover/provider.go#L90) them on by one by calling the `AutodiscoverBuilder` func.

Reviewer: ```suggestion
2. Metricbeat calls `NewAutodiscover` [function](https://github.com/elastic/beats/blob/master/metricbeat/beater/metricbeat.go#L183) which checks in the config for enabled providers and [builds](https://github.com/elastic/beats/blob/master/libbeat/autodiscover/provider.go#L90) them one by one, calling the `AutodiscoverBuilder` func.
```

> 3. Kubernetes `AutodiscoverBuilder` creates and returns a [Kubernetes Provider struct](https://github.com/elastic/beats/blob/master/libbeat/autodiscover/providers/kubernetes/kubernetes.go#L131) which is then added in a Autodiscover manager struct.

Reviewer: ```suggestion
3. Kubernetes `AutodiscoverBuilder` creates and returns a [Kubernetes Provider struct](https://github.com/elastic/beats/blob/master/libbeat/autodiscover/providers/kubernetes/kubernetes.go#L131) which is then added to an Autodiscover manager struct.
```

4. When unique is set to true [NewLeaderElectionManager](https://github.com/elastic/beats/blob/master/libbeat/autodiscover/providers/kubernetes/kubernetes.go#L141) is set as the eventManager of Kubernetes Provider.
4. Metricbeat [starts](https://github.com/elastic/beats/blob/master/metricbeat/beater/metricbeat.go#L249) the Autodiscover manager which starts for Kubernets provider the [leaderElectionManager](https://github.com/elastic/beats/blob/master/libbeat/autodiscover/providers/kubernetes/kubernetes.go#L326). Before starting the providers it also starts a worker for listening of events that will be published by the eventers of each provider.

> 5. OnStartedLeading is executed when the specific metricbeat instance gains the leader election lock. [StartLeading](startLeading) cretaes a bus event with `"start":    true,`  and publishes it. The template configurations is also added in this event.

Reviewer: ```suggestion
5. `OnStartedLeading` is executed when the specific metricbeat instance gains the leader election lock. [StartLeading](startLeading) cretaes a bus event with `"start":    true,`  and publishes it. The template configurations is also added in this event.
```

6. [Listener](https://github.com/elastic/beats/blob/master/libbeat/autodiscover/autodiscover.go#L140) of events get the published event and generates [configs](https://github.com/elastic/beats/blob/master/libbeat/autodiscover/autodiscover.go#L185) for it. Configs include the variables and settings from the template set by the user.
7. For each config the worker checks if it already [exists](https://github.com/elastic/beats/blob/master/libbeat/autodiscover/autodiscover.go#L225)(It was already handled). If at least of the configs of the event does not then the config is marked as `updated`.
8. The runners list get [reloaded](https://github.com/elastic/beats/blob/4b1f69923b3f2abbbf1860295fe5dbff7db3d63c/libbeat/cfgfile/list.go#L54). It is checked from the list of current runners if each config is handled by one of them.
   If no runner is handling that config a new runner will [start](https://github.com/elastic/beats/blob/4b1f69923b3f2abbbf1860295fe5dbff7db3d63c/libbeat/cfgfile/list.go#L107). If some runners are no longer needed will be removed.

> 9. Starting of a runner starts under the hood the metricbeat [module](https://github.com/elastic/beats/blob/4b1f69923b3f2abbbf1860295fe5dbff7db3d63c/metricbeat/mb/module/runner.go#L76) with the specific metricsets as they are set in the template.

Reviewer: Same happens for filebeat and hearbeat and not only for metricbeat. Maybe you can somehow make it more generic here?

### Autodiscover without LeaderElection

> When setting `unique: false` or not setting it at all the Leader Election mechanism is disabled. That way all the beat instances will enable the provided template. Or at least the ones that match a condition.

Reviewer: ```suggestion
When setting `unique: false` or not setting it at all the Leader Election mechanism is disabled. That way all Beat instances will enable the provided template. Or at least the ones that match a condition.
```

* [OnDelete](https://github.com/elastic/beats/blob/4b1f69923b3f2abbbf1860295fe5dbff7db3d63c/libbeat/autodiscover/providers/kubernetes/pod.go#L160) is called when there is a pod deletion and it ecexutes emit with `stop` flag. Emit adds the pod to a new event with ``stop: true`` and publishes it. Step 7 follows.
* [OnUpdate](https://github.com/elastic/beats/blob/4b1f69923b3f2abbbf1860295fe5dbff7db3d63c/libbeat/autodiscover/providers/kubernetes/pod.go#L146) is called when there is update in the an existing resource and it ecexutes at first emit with `stop` flag to generate a stop event and then with `start` flag to generate a start event.

## Hints Based Autodiscover

https://www.elastic.co/guide/en/beats/metricbeat/master/configuration-autodiscover-hints.html

> Metricbeat and Filebeat support autodiscover based on hints from the provider. The hints system looks for hints in Kubernetes Pod annotations which have the prefix co.elastic.metrics. As soon as the container starts, Metricbeat/Filebeat will check if it contains any hints and launch the proper config for it. Hints tell Metricbeat/Filebeat how to get metrics for the given container.

Reviewer: This also applies for Heartbeat.

9. Wether hints are enabled or not is part of the [Kubernetes Provider struct](https://github.com/elastic/beats/blob/master/libbeat/autodiscover/providers/kubernetes/kubernetes.go#L121) builders field.
10. GenerateHints[https://github.com/elastic/beats/blob/eff92354db783001880f4bade9f59942fca747ba/libbeat/autodiscover/builder/helper.go#L213] function looks into the event's annotations. A [hints map](https://github.com/elastic/beats/blob/eff92354db783001880f4bade9f59942fca747ba/libbeat/autodiscover/builder/helper.go#L226) is created with all hints and returned.
11. From those hints, configs are [created](https://github.com/elastic/beats/blob/master/libbeat/autodiscover/builder.go#L97) in the same form as in `Autodiscover without LeaderElection` step 8.
    They contain the same information as if they where set explicitly in the metricbeat configureation but actually derive from the pod annotations.
12. Those configs are then [added](https://github.com/elastic/beats/blob/master/libbeat/autodiscover/providers/kubernetes/kubernetes.go#L197) in the event and gets published.
13. The process after that is same as in `Autodiscover without LeaderElection` step 9 and onward.

> ##  Autodiscover provider in Elastic Agent

Reviewer: I would suggest moving this part under `elastic-agent` codebase so as to be close to the provider's code. Then you can cross-link it from libbeat's docs.

https://www.elastic.co/guide/en/fleet/master/dynamic-input-configuration.html#dynamic-providers

Currently Kubernetes dynamic provider can only be configured in [standalone](https://github.com/elastic/beats/blob/master/deploy/kubernetes/elastic-agent-standalone-kubernetes.yaml#L24) agent.
In fleet managed agent it is enabled by default with default values.

Template based autodiscover of Kubernetes resources is only supported in standalone mode as of now.
It is not part of the Kubernetes Integration yet.

Hints based autodiscover is not supported yet.

### Template based autodiscover

Example:

> As en example we will use gain redis module.

Reviewer: ```suggestion
As an example we will use gain redis module.
```

> As en example we will use gain redis module.

Reviewer: ```suggestion
As en example we will use again redis module.
```

> As en example we will use gain redis module.

Reviewer: ```suggestion
As en example we will use gain the `redis` module.
```

In agent.yml(configmap) an extra input block needs to be added.
```

>       # Add extra input blocks here, based on conditions

Reviewer: indentation?
